### PR TITLE
Add support for starting an SDM from a remote repo

### PR DESCRIPTION
**Pull request by cdupuis**
This allows to start SDMs with the following command:

```
$ atomist start --repository-url=git@github.com:cdupuis/minimal-sdm.git
```

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -221,13 +222,49 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
...
+                required: false,
+                default: "https://github.com/atomist-seeds/empty-sdm.git",
+            }, {
+                parameterName: "sha",
+                describe: "Git sha to checkout",
+                type: "string",
+                required: false,
+                default: "master",
+            }, {
+                parameterName: "file",
```
Comment (ddgenome) [inline index.ts]: Do you think "index" or "main" would be a better name?

File: index.ts [hunk 2]
```diff
@@ -221,13 +222,49 @@ function setupYargs(yargBuilder: yb.YargBuilder): void {
...
+                required: false,
             }],
-        handler: (argv: any) => cliCommand(() => start({
-            cwd: argv["change-dir"],
-            install: argv.install,
-            compile: argv.compile,
-            local: argv.local,
-        })),
+        handler: (argv: any) => cliCommand(() => {
+            if (!!argv["repository-url"]) {
```
Comment (ddgenome) [inline index.ts]: It seems odd that we diverge into two different implementations here. Plus, some of the command-line options only apply to one or the other. Perhaps starting from a repo should be its own command, something like "remote-start" or "start-repo".

File: lib/repositoryStart.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        local: false,
+        seedUrl: "https://github.com/atomist-seeds/empty-sdm.git",
+        compile: true,
+        install: true,
+        ...opts,
+    };
+
+    const gitUrl = gitUrlParse(optsToUse.cloneUrl);
+
+    const cwd = path.join(os.homedir(), ".atomist", "sdm", gitUrl.owner, gitUrl.name);
```
Comment (ddgenome) [inline lib/repositoryStart.ts]: Should this use the Atomist project directory, or whatever that is?

File: lib/repositoryStart.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    const seed = path.join(os.homedir(), ".atomist", "cache", `sdm-${guid().slice(0, 7)}`);
```
Comment (ddgenome) [inline lib/repositoryStart.ts]: Same.

File: lib/repositoryStart.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    // Copy over package.json, tsconfig.json in case they are missing
+    let packageJsonCreated = false;
+    try {
+        packageJsonCreated = await copyFiles(optsToUse, cwd, seed);
+    } catch (e) {
+        print.error(`Failed to copy seed files into clone: ${e.message}`);
+        return 15;
+    }
+
+    // Merge the provided package.json with the one from the seed as long there isn't a package-lock.json
```
Comment (ddgenome) [inline lib/repositoryStart.ts]: Why this logic? The assumption is if they have a package-lock.json then the Atomist dependencies must be included?

File: index.ts (hunk 1 above)
Comment (cdupuis) [reply index.ts]: yeah, I like `index`. Thanks

File: index.ts (hunk 2 above)
Comment (cdupuis) [reply index.ts]: True, I like the simplicity of one command. What if I merged the two implementations into one command and did the right thing in the implementation? 

That wouldn't solve the parameter dependencies. Can't you describe parameter dependencies with yargs. I'll check.

File: lib/repositoryStart.ts (hunk 3 above)
Comment (cdupuis) [reply lib/repositoryStart.ts]: You mean the `~/atomist` one? Not sure I want to put these into that directory. I contemplated that though and ended up not liking the idea.

File: lib/repositoryStart.ts (hunk 5 above)
Comment (cdupuis) [reply lib/repositoryStart.ts]: Yeah, I needed something to figure out when to stop merging in the package.json from the seed. I thought having a package-lock.json in the project is a good indication. 

Also had an implementation that used a special key in the package.json to trigger merging. But that felt like too much to ask.

Do you have a different suggestion?

Comment (ddgenome) [reply lib/repositoryStart.ts]: Why are you merging at all, to make sure the needed dependencies are in the package.json? Any other reason? Perhaps we should just be smarter about that. Check if the needed dependencies are there. If so, do nothing. If not, add them and remove the package-lock.json.

File: index.ts (hunk 2 above)
Comment (cdupuis) [reply index.ts]: I made the options depend on each other now. That clear this up.

**Comment by cdupuis** [comment]
Pull request auto merged by Atomist.

* 1 approved review by @ddgenome
* 2 successful checks

[atomist:generated] [auto-merge:on-approve]

**Pull request merged**